Mesh_3/include/CGAL/Mesh_3/Construct_initial_points_labeled_image.h

```diff
@@ -0,0 +1,299 @@
+      , tx(image->tx())
+      , ty(image->ty())
+      , tz(image->tz())
+      , xdim(image->xdim())
+      , ydim(image->ydim())
+      , zdim(image->zdim())
+  {}
+
+  Point operator()(const std::size_t i,
+                   const std::size_t j,
+                   const std::size_t k) const
+  {
+    double x = double(i) * vx + tx;
+    double y = double(j) * vy + ty;
+    double z = double(k) * vz + tz;
+
+    if (i == 0)              x += 1. / 6. * vx;
+    else if (i == xdim - 1)  x -= 1. / 6. * vx;
+    if (j == 0)              y += 1. / 6. * vy;
+    else if (j == ydim - 1)  y -= 1. / 6. * vy;
+    if (k == 0)              z += 1. / 6. * vz;
+    else if (k == zdim - 1)  z -= 1. / 6. * vz;
+
+    return Point(x, y, z);
+  }
+};
+
+} // end namespace internal
+} // end namespace Mesh_3
+
+/*!
+* \ingroup PkgMesh3Initializers
+*
+* Functor for initial points generation in labeled images.
+* This functor is a model of concept `InitialPointsGenerator`,
+* and thus can be passed to `CGAL::make_mesh_3` with the parameter
+* `CGAL::parameters::initial_points_generator()`
+*
+* \sa `CGAL::parameters::initial_points_generator()`
+* \sa `CGAL::make_mesh_3()`
+* \sa `CGAL::Construct_initial_points_gray_image`
+*/
+struct Construct_initial_points_labeled_image
+{
+  const CGAL::Image_3 & image;
+
+  Construct_initial_points_labeled_image(const CGAL::Image_3 & image_)
+      : image(image_)
+  { }
+
+   /*!
+   * \brief Constructs points by collecting them in all connected components.
+   * This guarantees to initialize them all.
+   *
+   * @tparam OutputIterator model of `OutputIterator`, containing points of type
+   * `std::tuple<MeshDomain::Point_3, int, MeshDomain::Index>`
+   * @tparam MeshDomain model of `MeshDomain_3`
+   * @tparam C3t3 model of `MeshComplex_3InTriangulation_3`
+   */
+  template <typename OutputIterator, typename MeshDomain, typename C3t3>
+  OutputIterator operator()(OutputIterator pts, const MeshDomain& domain, const C3t3& c3t3, int n = 20) const
+  {
+    CGAL_IMAGE_IO_CASE(image.image(), operator()(pts, domain, CGAL::Identity<Word>(), c3t3, n));
+    return pts;
+  }
+
+  /*!
+   * \brief Same as above, but a `TransformOperator` is used
+   *
+   * @tparam OutputIterator model of `OutputIterator`, containing points of type
+   * `std::tuple<MeshDomain::Point_3, int, MeshDomain::Index>`
+   * @tparam MeshDomain model of `MeshDomain_3`
+   * @tparam TransformOperator functor that transforms values of the image.
+   * It must provide the following type:<br>
+   * `result_type`<br>
+   * and the following operator:<br>
+   * `template<typename FT>`<br>
+   * `result_type operator()(FT v)`
+   * @tparam C3t3 model of `MeshComplex_3InTriangulation_3`
+   */
+  template <typename OutputIterator, typename MeshDomain, typename C3t3, typename TransformOperator>
+  OutputIterator operator()(OutputIterator pts, const MeshDomain& domain, TransformOperator transform, const C3t3& c3t3, int n = 20) const
+  {
```

Comment:
I do not get what is the semantic of `TransformOperator`. The worst is that I probably wrote this code myself, but I do not remember what I was thinking at that time.

Is there a use case somewhere in CGAL?

Reply:
In `master`, this functor is the one that enables to use `initialize_triangulation_from_labeled_image()` both for a labeled image (for which `Transform` is `Identity<Label>`) and a gray image (for which `Transform` virtually converts the gray levels into labels).

Maybe we should not document this version of `operator()` because it is an implementation detail of the two others.